app/src/main/java/com/sdp13epfl2021/projmag/database/UserProfileDatabase.kt

```diff
@@ -59,69 +65,63 @@ class UserProfileDatabase(
                             Gender.FEMALE.name -> Gender.FEMALE
                             else -> Gender.OTHER
                         }
-                        val phoneNumber = document["phoneNumber"] as? String
-                        val roleString = document["role"] as? String
-                        val sciperLong = document["sciper"]
 
-                        val role = when (roleString) {
+                        val role = when (document["role"] as? String) {
                             Role.TEACHER.name -> Role.TEACHER
                             Role.STUDENT.name -> Role.STUDENT
                             else -> Role.OTHER
                         }
 
-
-                        var age: Int = try {
-                            Integer.valueOf(ageLong.toString())
-                        } catch (excep: NumberFormatException) {
-                            0
-                        }
-
-                        var sciper: Int = try {
-                            Integer.valueOf(sciperLong.toString())
-                        } catch (exep: NumberFormatException) {
-                            0
-                        }
-
-
-                        when (val resProfile = ImmutableProfile.build(
-                            lastName ?: "null",
-                            firstName ?: "null",
-                            age,
-                            gender,
-                            sciper,
-                            phoneNumber ?: "null",
-                            role
-                        )) {
-                            is Success -> {
-                                onSuccess(resProfile.value)
-                            }
-                            is Failure -> {
-                                Log.d(TAG, "Failure reason : ${resProfile.reason}")
+                        if (firstName != null && lastName != null && age != null && sciper != null && phoneNumber != null) {
+                            when (val resProfile = ImmutableProfile.build(
+                                lastName,
+                                firstName,
+                                age,
+                                gender,
+                                sciper,
+                                phoneNumber,
+                                role
+                            )) {
+                                is Success -> {
+                                    onSuccess(resProfile.value)
+                                }
+                                is Failure -> {
+                                    onFailure(Exception("Failure reason : ${resProfile.reason}"))
+                                }
                             }
+                        } else {
+                            onFailure(Exception("At least one of the following fields is null: firstName = $firstName, lastName = $lastName, age = $age, sciper = $sciper, phoneNumber = $phoneNumber."))
```

Comment:
Extracting messages to strings.xml could be better.
getString can be used with parameters (such as %s text, %d for number)

Reply:
strings.xml is main usage is to translate between the different languages the app will be deployed in, so in this specific case it does not make sense to add it to strings.xml.